### PR TITLE
[RFC4_dev] Add networking capabilities

### PR DESCRIPTION
**Pull request by rouault**
This is a continuation of #1790. Only the commits starting with "Rename PJ_CONTEXT::fileapi member as fileapi_legacy" are specific to this pull request.

Summary:
- Add a optional libcurl dependency (enabled on all CI targets, except the mingw32 one)
- Allow alternate network callbacks to be plugged in.
- Networking only enabled if allowed with PROJ_NETWORK=ON, proj_context_set_enable_network() or through the proj.ini configuration file
- Network access is done by chunks of multiples of 16 KB
- In-memory cache of downloaded chunks
- Default network endpoint is https://cdn.proj.org (defined in proj.ini). Can also be configured with PROJ_NETWORK_ENDPOINT or proj_context_set_url_endpoint()
- When network is enabled, grid filenames starting with http[s]:// are allowed
- When network is enabled, pj_create_crs_to_crs[_from_pj] will consider that grids in `grid_alternatives` can be used even if not available locally.  With lower level API, can be done with `proj_operation_factory_context_set_grid_availability_use(..., PROJ_GRID_AVAILABILITY_KNOWN_AVAILABLE)`
- When network is enabled, grids referenced by a short name will be looked up remotely if not available locally. e.g grids=egm08_25.gtx will be substituted by grids=https://cdn.proj.org/egm08_25.tif
- Adds a test/unit/test_network.cpp file

Not yet implemented: on-disk cache of downloaded chunks

### CODE REVIEW

File: docs/source/apps/cs2cs.rst [hunk 1]
```diff
@@ -175,6 +175,14 @@ normally be in DMS format (use ``-f %.12f`` for decimal degrees with 12 decimal
 places), while projected (cartesian) coordinates will be in linear
 (meter, feet) units.
 
+Use of remote grids
```
Comment (kbevers) [inline docs/source/apps/cs2cs.rst:178]: Perhaps this should also be added to `docs/source/apps/cct.rst`. I assume the same applies.

**Review by kbevers — APPROVED** [review]
I am not in a position to thoroughly review all of the technical bits in this PR but the overall structure looks good to me. The reworking of the filemanager looks like a big improvement. I wonder if it can easily be exposed via `proj.h`? If so, we would have a replacement for the old file accessing API and  #866 can be dealt with properly.

File: docs/source/apps/cs2cs.rst (hunk 1 above)
Comment (rouault) [reply docs/source/apps/cs2cs.rst:178]: done in https://github.com/OSGeo/PROJ/pull/1839/commits/4d21a242a10cf0e3c5b9503a9d0e8013ed4afd78

**Comment by rouault** [comment]
>  I wonder if it can easily be exposed via `proj.h`

Done per https://github.com/OSGeo/PROJ/pull/1839/commits/55625528a17f7fe36d4100f23497ea77b95d39a6

**Comment by kbevers** [comment]
> Done per 5562552

Thank you. The public functions needs documentation. I can do it if you want, I just need to know which branch to commit on.

**Comment by rouault** [comment]
> The public functions needs documentation

As they already have Doxygen doc, this is addressed by https://github.com/OSGeo/PROJ/pull/1839/commits/c40ee9dfbbac32ebf3daa1b9edca64a247d4ad53 that just uses Breathe to include it in the appropriate sections

**Comment by kbevers** [comment]
> As they already have Doxygen doc, this is addressed by c40ee9d that just uses Breathe to include it in the appropriate sections

Perfect!

**Pull request merged**